[JENKINS-76126] Pipeline Maven Integration plugin fails with FindbugsAnalysisPublisher is missing its descriptor exception

<!-- ### Imported Jira references for easier searching -->
<!-- [jira_issue_key=JENKINS-76126] -->
<!-- [jira_issue_type=Bug] -->
<!-- [reporter=denys_mazhar_quant] -->
<!-- [author=denys_mazhar_quant] -->
<!-- [jira_component=pipeline-maven-plugin] -->
<!-- [jira_issues_importer_version=1.4.1] -->

<p>Hi team. I updated the pipeline maven integration plugin from version <tt>1508.v347c4b_692202</tt> to <tt>1567.vb_2c3a_2116860</tt> and our builds started fail with an error:<br/>
org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 37fc4de5-73c5-47ea-944d-664a08a0399b<br/>
java.lang.AssertionError: class org.jenkinsci.plugins.pipeline.maven.publishers.FindbugsAnalysisPublisher is missing its descriptor<br/>
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1554)<br/>
	at hudson.model.AbstractDescribableImpl.getDescriptor(AbstractDescribableImpl.java:43)<br/>
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.MavenPublisher.getDescriptor(MavenPublisher.java:47)<br/>
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.MavenPublisherStrategy$1.buildPublishersList(MavenPublisherStrategy.java:66)<br/>
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.MavenSpyLogProcessor.processMavenSpyLogs(MavenSpyLogProcessor.java:144)<br/>
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.WithMavenStepExecution2$WithMavenStepExecutionCallBack.finished(WithMavenStepExecution2.java:1365)<br/>
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution$TailCall.lambda$onSuccess$0(GeneralNonBlockingStepExecution.java:141)<br/>
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution.lambda$run$0(GeneralNonBlockingStepExecution.java:77)<br/>
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)<br/>
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)<br/>
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)<br/>
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)<br/>
	at java.base/java.lang.Thread.run(Thread.java:840)<br/>
We have no findbugs plugin installed, but com.google.code.findbugs maven plugin's used. We do not really need finbugs report published in Jenkins, so previously plugin was just showing warning messages:<br/>
<b>16:24:18</b>  Recording test results*16:24:26*  <span class="error">&#91;Checks API&#93;</span> No suitable checks publisher found.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Jenkins <a href="https://wiki.jenkins-ci.org/display/JENKINS/FindBugs+Plugin" class="external-link" target="_blank" rel="nofollow noopener">FindBugs Plugin</a> not found, don't display org.codehaus.mojo:findbugs-maven-plugin:findbugs results in pipeline screen.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> jgivenPublisher - Jenkins <a href="https://wiki.jenkins.io/display/JENKINS/JGiven+Plugin" class="external-link" target="_blank" rel="nofollow noopener">JGiven Plugin</a> not found, do not archive jgiven reports.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Jenkins <a href="https://wiki.jenkins-ci.org/display/JENKINS/Task+Scanner+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Task Scanner Plugin</a> not found, don't display results of source code scanning for 'TODO' and 'FIXME' in pipeline screen.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Publishers: Pipeline Graph Publisher: 40 ms, Invoker Publisher: 4 ms, Junit Publisher: 8114 ms, Concordion Publisher: 2 ms, Dependencies Fingerprint Publisher: 799 ms, Findbugs Publisher: 1 ms, JGiven Publisher: 1 ms, SpotBugs Publisher: 1 ms, Open Task Scanner Publisher: 1 ms*16:24:27*  <span class="error">&#91;withMaven&#93;</span> JacocoPublisher is no longer implicitly used with `withMaven` step.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Jacoco plugin has been deprecated and should not be used anymore.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Usage of Coverage plugin is recommended.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> See <a href="https://plugins.jenkins.io/coverage/#plugin-content-pipeline-example" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/coverage/#plugin-content-pipeline-example</a>.<br/>
Rolling back to the version <tt>1567.vb_2c3a_2116860</tt> fixed an issue.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=denys_mazhar_quant">denys_mazhar_quant</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-76126" target="_blank">Pipeline Maven Integration plugin fails with FindbugsAnalysisPublisher is missing its descriptor exception</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: pipeline-maven-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 0
<li><b>watchers</b>: 3
<li><b>imported</b>: 20251211-223321
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Hi team. I updated the pipeline maven integration plugin from version <tt>1508.v347c4b_692202</tt> to <tt>1567.vb_2c3a_2116860</tt> and our builds started fail with an error:
org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 37fc4de5-73c5-47ea-944d-664a08a0399b
java.lang.AssertionError: class org.jenkinsci.plugins.pipeline.maven.publishers.FindbugsAnalysisPublisher is missing its descriptor
	at jenkins.model.Jenkins.getDescriptorOrDie(Jenkins.java:1554)
	at hudson.model.AbstractDescribableImpl.getDescriptor(AbstractDescribableImpl.java:43)
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.MavenPublisher.getDescriptor(MavenPublisher.java:47)
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.MavenPublisherStrategy$1.buildPublishersList(MavenPublisherStrategy.java:66)
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.MavenSpyLogProcessor.processMavenSpyLogs(MavenSpyLogProcessor.java:144)
	at PluginClassLoader for pipeline-maven//org.jenkinsci.plugins.pipeline.maven.WithMavenStepExecution2$WithMavenStepExecutionCallBack.finished(WithMavenStepExecution2.java:1365)
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution$TailCall.lambda$onSuccess$0(GeneralNonBlockingStepExecution.java:141)
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution.lambda$run$0(GeneralNonBlockingStepExecution.java:77)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
We have no findbugs plugin installed, but com.google.code.findbugs maven plugin's used. We do not really need finbugs report published in Jenkins, so previously plugin was just showing warning messages:
<b>16:24:18</b>  Recording test results*16:24:26*  <span class="error">&#91;Checks API&#93;</span> No suitable checks publisher found.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Jenkins <a href="https://wiki.jenkins-ci.org/display/JENKINS/FindBugs+Plugin" class="external-link" target="_blank" rel="nofollow noopener">FindBugs Plugin</a> not found, don't display org.codehaus.mojo:findbugs-maven-plugin:findbugs results in pipeline screen.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> jgivenPublisher - Jenkins <a href="https://wiki.jenkins.io/display/JENKINS/JGiven+Plugin" class="external-link" target="_blank" rel="nofollow noopener">JGiven Plugin</a> not found, do not archive jgiven reports.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Jenkins <a href="https://wiki.jenkins-ci.org/display/JENKINS/Task+Scanner+Plugin" class="external-link" target="_blank" rel="nofollow noopener">Task Scanner Plugin</a> not found, don't display results of source code scanning for 'TODO' and 'FIXME' in pipeline screen.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Publishers: Pipeline Graph Publisher: 40 ms, Invoker Publisher: 4 ms, Junit Publisher: 8114 ms, Concordion Publisher: 2 ms, Dependencies Fingerprint Publisher: 799 ms, Findbugs Publisher: 1 ms, JGiven Publisher: 1 ms, SpotBugs Publisher: 1 ms, Open Task Scanner Publisher: 1 ms*16:24:27*  <span class="error">&#91;withMaven&#93;</span> JacocoPublisher is no longer implicitly used with `withMaven` step.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Jacoco plugin has been deprecated and should not be used anymore.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> Usage of Coverage plugin is recommended.<b>16:24:27</b>  <span class="error">&#91;withMaven&#93;</span> See <a href="https://plugins.jenkins.io/coverage/#plugin-content-pipeline-example" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/coverage/#plugin-content-pipeline-example</a>.
Rolling back to the version <tt>1567.vb_2c3a_2116860</tt> fixed an issue.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins version: 2.492.3<br/>
Pipeline Maven Integration plugin version: 1567.vb_2c3a_2116860<br/>
```
</details>

--- a/pipeline-maven/src/main/java/org/jenkinsci/plugins/pipeline/maven/MavenPublisherStrategy.java
+++ b/pipeline-maven/src/main/java/org/jenkinsci/plugins/pipeline/maven/MavenPublisherStrategy.java
@@ -47,7 +47,22 @@ public enum MavenPublisherStrategy {
                 if (mavenPublisher == null) {
                     // skip null publisher options injected by Jenkins pipeline, probably caused by a missing plugin
                 } else {
-                    configuredPublishersById.put(mavenPublisher.getDescriptor().getId(), mavenPublisher);
+                    try {
+                        configuredPublishersById.put(
+                                mavenPublisher.getDescriptor().getId(), mavenPublisher);
+                    } catch (AssertionError e) {
+                        // Descriptor missing — plugin uninstalled/deprecated
+                        listener.getLogger()
+                                .println("[withMaven] WARNING: Skipping publisher '"
+                                        + mavenPublisher.getClass().getSimpleName()
+                                        + "' as its descriptor is missing. "
+                                        + "The associated plugin may be uninstalled or deprecated.");
+                        LOGGER.log(
+                                Level.WARNING,
+                                "Skipping publisher with missing descriptor: "
+                                        + mavenPublisher.getClass().getName(),
+                                e);
+                    }
                 }
             }
 
@@ -62,8 +77,21 @@ public enum MavenPublisherStrategy {
                 globallyConfiguredPublishers = Collections.emptyList();
             }
             for (MavenPublisher mavenPublisher : globallyConfiguredPublishers) {
-                globallyConfiguredPublishersById.put(
-                        mavenPublisher.getDescriptor().getId(), mavenPublisher);
+                try {
+                    globallyConfiguredPublishersById.put(
+                            mavenPublisher.getDescriptor().getId(), mavenPublisher);
+                } catch (AssertionError e) {
+                    listener.getLogger()
+                            .println("[withMaven] WARNING: Skipping globally configured publisher '"
+                                    + mavenPublisher.getClass().getSimpleName()
+                                    + "' as its descriptor is missing. "
+                                    + "The associated plugin may be uninstalled or deprecated.");
+                    LOGGER.log(
+                            Level.WARNING,
+                            "Skipping globally configured publisher with missing descriptor: "
+                                    + mavenPublisher.getClass().getName(),
+                            e);
+                }
             }
 
             // mavenPublisher.descriptor.id -> mavenPublisher

--- a/pipeline-maven/src/test/java/org/jenkinsci/plugins/pipeline/maven/MavenPublisherStrategyTest.java
+++ b/pipeline-maven/src/test/java/org/jenkinsci/plugins/pipeline/maven/MavenPublisherStrategyTest.java
@@ -1,6 +1,8 @@
 package org.jenkinsci.plugins.pipeline.maven;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 import hudson.util.StreamTaskListener;
 import java.io.ByteArrayOutputStream;
@@ -56,5 +58,19 @@ public class MavenPublisherStrategyTest {
         assertThat(reportersByDescriptorId).containsKey(new MavenLinkerPublisher2.DescriptorImpl().getId());
         assertThat(reportersByDescriptorId).containsKey(new PipelineGraphPublisher.DescriptorImpl().getId());
         assertThat(reportersByDescriptorId).containsKey(new CoveragePublisher.DescriptorImpl().getId());
+    }
+
+    @Test
+    public void buildPublishersListShouldSkipPublisherWithMissingDescriptor(JenkinsRule r) throws Exception {
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+        MavenPublisher brokenPublisher = mock(MavenPublisher.class);
+        when(brokenPublisher.getDescriptor()).thenThrow(new AssertionError("Descriptor is missing!"));
+
+        List<MavenPublisher> result = MavenPublisherStrategy.IMPLICIT.buildPublishersList(
+                Collections.singletonList(brokenPublisher), new StreamTaskListener(baos));
+
+        assertThat(baos.toString()).contains("WARNING");
     }
 }

